src/eos-update-server.c

```diff
@@ -0,0 +1,863 @@
+  return TRUE;
+}
+
+static gboolean
+serve_remote_goption (const gchar *option_name,
+                      const gchar *value,
+                      gpointer options_ptr,
+                      GError **error)
+{
+  Options* options = options_ptr;
+  g_autoptr(GError) local_error = NULL;
+  g_autofree gchar *remote = NULL;
+  g_autofree gchar *test_refspec = g_strdup_printf ("%s:test", value);
+
+  if (g_strcmp0 (option_name, "--serve-remote") != 0 && g_strcmp0 (option_name, "-r") != 0)
+    {
+      g_set_error (error, G_OPTION_ERROR, G_OPTION_ERROR_FAILED,
+                   "Wrong option %s to parse, should not happen", option_name);
+      return FALSE;
+    }
+  if (!ostree_parse_refspec (test_refspec, &remote, NULL, NULL) ||
+      g_strcmp0 (value, remote) != 0)
+    {
+      g_set_error (error, G_OPTION_ERROR, G_OPTION_ERROR_FAILED,
+                   "Badly-formed remote name %s", value);
+      return FALSE;
+    }
+
+  g_free (options->served_remote);
+  options->served_remote = g_steal_pointer (&remote);
+  return TRUE;
+}
+
+static gboolean
+options_init (Options *options,
+              int *argc,
+              gchar ***argv,
+              GError **error)
+{
+  g_autoptr(GOptionContext) context = NULL;
+  g_autoptr(GOptionGroup) group = NULL;
+  GOptionEntry entries[] = {
+    { "local-port", 'p', G_OPTION_FLAG_NONE, G_OPTION_ARG_CALLBACK, local_port_goption, "Local port number (0 < N < 65536)", "N" },
+    { "timeout", 't', G_OPTION_FLAG_NONE, G_OPTION_ARG_INT, &options->timeout_seconds, "Time in seconds of inactivity allowed before quitting (zero or less means no timeout), default 5 seconds", "N" },
+    { "serve-remote", 'r', G_OPTION_FLAG_NONE, G_OPTION_ARG_CALLBACK, serve_remote_goption, "Which remote should be served, default eos", "NAME" },
+    { NULL }
+  };
+
+  context = g_option_context_new ("Endless OSTree server");
+  group = g_option_group_new (NULL, NULL, NULL, options, NULL);
+  g_option_group_add_entries (group, entries);
+  g_option_context_set_main_group (context, g_steal_pointer (&group));
+
+  memset (options, 0, sizeof (*options));
+  options->timeout_seconds = 5;
+  options->served_remote = g_strdup ("eos");
+
+  return g_option_context_parse (context, argc, argv, error);
+};
+
+static void
+options_clear (Options *options)
+{
+  if (options == NULL)
+    return;
+
+  options->local_port = 0;
+  options->timeout_seconds = 0;
+  g_clear_pointer (&options->served_remote, g_free);
+}
+
+G_DEFINE_AUTO_CLEANUP_CLEAR_FUNC (Options, options_clear)
+
+static void
+clear_source (guint *id)
+{
+  guint source_id;
+
+  if (id == NULL || *id == 0)
+    return;
+  source_id = *id;
+  *id = 0;
+
+  g_source_remove (source_id);
+}
+
+typedef struct Data Data;
+
+struct Data
+{
+  OstreeRepo *repo;
+  GMainLoop *loop;
+  SoupServer *server;
+
+  gint timeout_seconds;
+  guint timeout_id;
+
+  gchar *remote_name;
+
+  gchar *cached_repo_root;
+  GBytes *cached_config;
+
+  guint async_requests_pending;
+};
```

Comment:
In a followup PR I made some changes here that made that struct smaller, so I renamed that to `TimeoutData`.